Workpiece tracking and intermediate scoring

MostafaGomaa opened the pull request:
The new feature make use of the barcode scanners newly installed at each machine, to keep track of the production processes performed on a workpiece, allowing us to finally award points for intermediate production steps as soon they are performed. The awarding criterion is described by the [rcll-rulebook](https://github.com/robocup-logistics/rcll-rulebook). 

This PR implements #15, fixes #24 and fixes #37.
 
For workpiece tracking to function properly, we need to ensure that all barcode readers at all machines function properly (especially, BS and RS).  A failing Barcode reader at CS won't hurt at all and at DS will cause us some damage that could be recovered from in some cases.  If a barcode reader fails at RS or BS the workpiece could be assumed to track the wrong order. There will be no means to rectify that error till the workpiece is delivered.  in other words,  It is NOT safe to assume that the tracking will function properly if one or more of the readers are broken. In this scenario, we recommend disabling the workpiece-tracking feature.


### Approach
- When workpiece tracking is enabled, every operation done at a machine are recorded in the fact product-processed.
- The operations are linked to a workpiece that was read at the machine.
- Each operation needs to be confirmed, that it describes the workpiece
--  BS operations are confirmed with the base color as soon as linked the workpiece.
--  RS operations are confirmed with the base color as soon as linked the workpiece.
--  CS operations are confirmed upon delivery confirmation as there is no way to know the cap color.
--  DS operations are confirmed by the referee.
- The immediate score is given for the confirmed operation of a workpiece that tracks a specific order (according to the scoring scheme).
- Upon delivery confirmation by the referee, Mounting-Cap and Delivery scores are awarded.


### Tesing 
to test the workpiece tracking feature in mockup mode:
- Enable workpiece-tracking in config.yaml
- Enable fail-safe if you wish to test the feature
- All you need to do is to send a fake a barcode reading event on a machine at any point before the machine finishes the operation (ie, before it goes to WAIT-IDLE). This can be done using the rcll-workpiece command line tool in /bin.
for example, `./rcll-workpiece send 1232 C-BS `
- Of course, you can simulate failures in barcode readers by not sending the events at a one or more machines
- To show what workpiece are currently in the system use 
 `./rcll-workpiece list `

### Usage Notes
**Fail-safe enabling** in the config will cause workpiece-tracking to be shut down upon 
 - a failed read at a BS|RS, ie, the operation finished at the machine before a barcode was read (once the machine has passed WAIT-IDLE).
- a failed read at a DS where the confirmed delivery could not be link to a fitting unique workpiece that finished its Cap operation.

**Fail-safe disabling** in config could have the following implications
- a failed barcode read at RS| BS,  could cause the workpiece to be assigned to track the wrong order. The only stage we can recover from this, is upon 'positive' delivery confirmation by the referee. Then, the workpiece is verified against this ground truth.  In case, the wrongly tracked workpiece was never/ wrongly delivered, we won't get a chance to rectify this tracking failure.  This could cause us to award or deny points for the wrong operation or by relating to the wrong order.
- It might still be a good idea to test the tracking with the fail-safe disabled as to test as much of the pipeline as possible without having to restart a new test game on every failure.

**Automatic failure-recovery**
 -  recovery from a failed read at a DS will be attempted upon referee delivery confirmation. An attempt will be made to link the delivery operation to a fitting workpiece if a unique exists which finished its cap operation.
-  failed read happening at CS  'only' ,  will not cause harm to the tracking as it will be rectified, later upon delivery confirmation. Yet, the following failure at DS will no longer be able to recover if CS read had failed already.
- Order ambiguity caused by the similarity of an order's prefix to another is corrected as soon as the information is available. Any score given for the wrong reason would be removed.
-  A workpiece will be verified upon its delivery confirmation.  When a workpiece is delivered, referee confirmation will be used as ground truth for what this workpiece should look like. Any inconsistencies between the workpiece and the order, will be rectified. This ensures that we always only award the points of building the product that has been delivered by this workpiece. Any points given for a wrongly tracked operation will be removed. 

**Failure recovery by human intervention (referee)**
- When a scanner fails to read for any reason at any machine, the referee could prevent this failure to have an effect on the tracking, if she/he manages to re-scan the workpiece before the machine goes to WAIT-IDLE. In case the machine was already at ready-at-output, a risky attempt could be to keep a finger at the output sensor while the workpiece is rescanned then returned to its place. It's necessary that this should be attempted skillfully without causing the machine to break or force the workpiece to be retrieved.
- A special case that is particularly useful is attempting to rescan when a scanner at DS fails. A workpiece could be re-scanned at any point before another delivery is attempted at the same DS. The workpiece should be safely linked to the last operation (even if DS had already IDLED). Needless to say, this needs to happen before the delivery is confirmed by referee, otherwise its too late and only the automatic recovery will be attempted.

morxa commented:
Can you rebase this please? So it doesn't contain the changes from #26, or wait for that PR to be merged and then rebase.

MostafaGomaa commented:
This is still work in progress, the main handling is there and tracking works ( if all barcodes perfectly works). 
So, I would still appreciate your feedback as to its more or less ready!!  (except the confirmation of delivery in order.clp)

I still want to apply some changes to order.clp t to at recover gracefully from a failed tracking at one of the MSPs, when delivery is confirmed.  (as we atm, have yet no idea how often or why could the readers fail )

I will rebase it once the #26 is in master as I am touching the same production rule and I was testing a complete rebase

MostafaGomaa commented:
This PR is finally ready again for reviews, looking forward :)

MostafaGomaa commented:
New 'fail-safe' feature:
I have added a new configurable fail-safe feature that would gracefully disable workpiece tracking when a barcode reader fails to scan at a machine. The severity of the effect of that failure depends on the machine type and if some mild recovery attempts are successful. When the effect is considered to affect the soundness of tracking, workpiece tracking is disabled and unfinished workpieces and their operations retracted. Generally, failure at BS|RS|DS triggers the fail-safe (except for DS we attempt to recover before we fail-safe)

I updated the PR Description post accordingly with more details

morxa commented on cfg/config.yaml:
```diff
@@ -155,6 +155,18 @@ llsfrb:
+  #                        In simulation this might be tedious and unnecessary.
+  #                        'Automatic' confirmation grants all delivery requests.
+  referee-confirmation:
+    delivery: manual
```
Please make this a boolean, e.g., `auto-confirm-delivery: true` (no need for nesting)

(Also, I would have preferred to have this in a separate PR, this has nothing to do with this PR)

MostafaGomaa replied:
I removed it from this PR, I will file another PR with your suggestion once #30 is merged into master.

The reason I had a nesting though, is that I think later might be useful to also specify how do we confirm a cap mount. Whether it should always be granted, confirmed by referee or confirmed on delivery. At the moment, we do not award the cap operation in its time due to the fact we can not tell if it was the correct cap color. Atm, the delivery confirmation is also used to determine the cap color correctness

morxa commented:
Sorry that the review took so long, #39 blocked me from testing for quite a while.

MostafaGomaa commented:
> Sorry that the review took so long, #39 blocked me from testing for quite a while.

Thanks for the review, indeed #39 is a bit annoying, I had not had the time to look-into what causes it. There is some tricks I do to get around for now. I will post that in the issue!

I will process your comment in the next hours, thank!

morxa commented:
> > Sorry that the review took so long, #39 blocked me from testing for quite a while.
> 
> Thanks for the review, indeed #39 is a bit annoying, I had not had the time to look-into what causes it. There is some tricks I do to get around for now. I will post that in the issue!

No need, I have a workaround, but I was trying to actually fix the issue.

MostafaGomaa commented on src/games/rcll/net.clp:
```diff
@@ -274,12 +275,80 @@
+
+	(printout t "Got " ?id " at " ?at-machine-str " visible? " ?visible crlf)
+
+  (assert (mps-status-feedback (sym-cat ?at-machine-str) BARCODE ?id))
```
Done :)

MostafaGomaa commented on src/games/rcll/net.clp:
```diff
@@ -274,12 +275,80 @@
+  (do-for-all-facts ((?client network-client)) (not ?client:is-slave)
+    (pb-send ?client:id ?wi))
+  (pb-destroy ?wi)
+)
```
Good question, so that was a part of the implementation by @timn that I left in place. I think that is an aspect of tracking that we might still need to discuss the reasoning behind it. Options we have
1- Send to all info to all clients, allowing game theoretic reasoning and full knowledge of what the other team is doing
2- Send only team-specific workpiece info to each client on the private channel allowing them for partial reasoning about the success of their operations. 
3- Do not send any workpiece information and only use it internally for tracking

1 and 2 could potentially have the disadvantage of influencing the team's reasoning with inaccurate information if tracking failed for any reason (for example, a failed reader at an mps)

I think for now we should conceal the workpiece information (pick 3).
What do you think?

MostafaGomaa replied:
I made it configurable!

MostafaGomaa commented on src/games/rcll/workpieces.clp:
```diff
@@ -0,0 +1,373 @@
+   This has a very specific assumption about the ring colors. We assume that
+   the color of the first ring fully determines the complexity class. Furthermore
+   we assume that there is only a single order of complexity >0 each. Hence, knowing
+   the first ring will immediately determine the order of the workpiece."
```
That is a good Idea but!
As a matter of fact, this assumption doesn't even completely hold anymore. Which makes this documentation even more wrong, thanks for pointing it out! 
I have noticed 2 C1 identical orders. This assumption was a part of the old code I extended. To some extent I have adapted to not having this assumption anymore by having the following assumptions.
- The orders can be resigened and re-assigened to a workpiece at anypoint when its needed.
- Points calculated and given for a resigend order are recalculated when assigened a new.
- Workpieces are always verified against the confirmed delivery and rectifed if needed.
(please, keep an eye on those during testing)

Yet, A recent test case that your review had sparked makes me reconsider ensuring the "unique order prefix assumption". That is, related to the allowance of scoring on identical operations for the identical orders. atm. the ambiouty could lead us to prevent a team of scoring points inteded for an order identical to the one we are tracking.
Senario:
- Order 2 and 8 are identical C1 (b:BLACK|r1:BLUE|c:SILVER)
- a team produces 2 and 8 at the same time.
- we assume both workpieces to track order 8.
We will only award the mount ring blue once (since order 8 quantitiy is 1). If we deliver this tracking error should be rectified and the points intended for order 2 will be given. The issue will be if they did not delivery the order 2.

I need to think about if we want to ensure the "unique prefix assumption" (for anyother reasons) or simply adapt that case to not having it!.

what do you think?

MostafaGomaa replied:
I took a closer look into this,
Violation of the assumption could lead us to many corner cases that tracking needs to make sure of actively correcting. The ambiguity generated is simply too annoying and sometimes there is actually no way to treat it till delivery is attempted by the team (which we already handle atm, our tracking error is only a problem when 2 orders with identical rings are not delivered)

The main issue was that we have 2 C1 orders posted, this violates the last part of the assumption
```
Furthermore, we assume that there is only a single order of complexity >0 each.
```

MostafaGomaa commented:
Thanks for the review!

> There is one major issue: I we do the steps for an order multiple times, we get points for
> those steps multiple times. The rules clearly state that this shouldn't be the case:

Indeed, this should already be the case!
 I have tried to regenerate this but I could not score for the same operation again. Can you please tell me how you generated this or send me the log file. I will run some more tests from my side and see what I can run into. One thing that comes to mind though, could it be that you had 2 identical C1 orders? (I just found out this is possible).

In any case,  the testing for this sparked a use case that would prevent us from given ring intermediate points if 2 orders are identical. (for details look at the reply below about the "unique prefix assumption").
 
> Also, I find the output a bit confusing:
> 
>     * after preparing C-CS1 for RETRIEVE_CAP and `./rcll-workpiece send 1 C-CS1`:
>       ```
>       10:45:50.281311 C: Got 1 at C-CS1 visible? FALSE
>       10:45:50.281436 C: Workpiece 1: at C-CS1, availabe!
>       10:45:50.281516 C: Invalid workpiece ID 1. Cannot determine base color.
>       ```
>       
>       
>       
>       * Why visible `FALSE`?
This is indeed confusing! I replaced it with a better print.  I had forgotten to remove the interpreting of the "visibility" field in the msg after I changed the semantics of the field to express the "visibility history".

>       * The invalid workpiece error message is distracting, why is this an error?
This is based on an assumption implemented by @timn at the branch base I extended.  The assumption is that the workpiece ID determines the color of the BASE. You can find the function that  interprets the ranges at the top of the workpiece.clp. 

Since this assumption was already made, I could see a benefit or tow for it, Specially, if this was already considered while printing the codes. Benefits I could see for the assumption.
- A way to find out if the base station dispense a wrong base (human or machine error)
- If we want to know the color of an non-dispensed base, at shelf or from storage (no current use-case for that but could be useful in the future)
- In simulation, the color of the spawned base could determine its ID
anyway, as I did know the real reason for the assumption and its clearly intentionally there,  I decided to keep the assumption as a first stage to determine the base color and correct the base-color when the information is available. 
>     * similarly, at the BS:
>       ```
>       10:47:08.241271 C: Got 2 at C-BS visible? FALSE
>       10:47:08.241398 C: Workpiece 2: at C-BS, availabe!
>       10:47:08.241499 C: Invalid workpiece ID 2. Cannot determine base color.
>       10:47:08.241631 C: Workpiece 2: at C-BS, processed
>       10:47:08.241706 C: Workpiece correction [BASE_INVALID->BASE_BLACK]
>       10:47:08.241880 C: Workpiece 2: order assigned 1
>       ```
This is again becasue the allowed ranges are a 4 digit integer and you sent a single digit. 
> 
> Please also see my other in-line comments. Good work though!
Thanks! please, have another look and let me know what you think about the open questions

MostafaGomaa commented on src/games/rcll/workpieces.clp:
```diff
@@ -0,0 +1,373 @@
+  =>
+  (printout t "Workpiece " ?id ": at " ?m-name ", processed"crlf)
+  (if (neq ?bs-color ?base-color)
+    then (printout t "Workpiece correction ["
```
I have detailed about this point in the reply for your first review comment. 

Briefly, The workpiece ID is assumed to have 4 digit coded ranges that determine their colors. The Idea here is that an ID should somewhat already determine the color. The correction is done when the color based on ID does not match the dispensed base (ie, the requested color by the prepare msg)

MostafaGomaa commented:
Ready for another look :)

morxa commented:
> Thanks for the review!
> 
> > There is one major issue: I we do the steps for an order multiple times, we get points for
> > those steps multiple times. The rules clearly state that this shouldn't be the case:
> 
> Indeed, this should already be the case!
> I have tried to regenerate this but I could not score for the same operation again. Can you please tell me how you generated this or send me the log file. I will run some more tests from my side and see what I can run into. One thing that comes to mind though, could it be that you had 2 identical C1 orders? (I just found out this is possible).

Yes, true, there were two orders!

> 
> > ```
> >   * The invalid workpiece error message is distracting, why is this an error?
> > ```
> 
> This is based on an assumption implemented by @timn at the branch base I extended. The assumption is that the workpiece ID determines the color of the BASE. You can find the function that interprets the ranges at the top of the workpiece.clp.

I see, so I was just using the wrong ID format?

MostafaGomaa commented:
Yup!
The workpiece Ids are 4 digit colour coded ranges.

MostafaGomaa commented:
Cool, Sounds good!
 Looking forward for your feedback on any issues :)

Lots of luck and safe travels

morxa commented:
Thanks for the great work!

MostafaGomaa commented:
Woo :D , Thanks for the reviews and testing
Good Luck!
